File: packages/sdk-typescript/test/attachments.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  UBAG_ATTACHMENT_MAX_FILE_BYTES,
  UBAG_ATTACHMENT_MAX_MANIFEST_FILES,
  UbagClient,
} from "../dist/index.js";

const request = {
  client: {},
  job: { target: "mock", command_type: "chat.prompt", input: { prompt: "inspect" } },
};

const response = (body) =>
  new Response(JSON.stringify(body), {
    status: 202,
    headers: { "content-type": "application/json" },
  });

test("createJobWithAttachments uploads every key-reference body", async () => {
  const seen = [];
  const client = new UbagClient({
    baseUrl: "https://gateway.example",
    fetch: async (url, init) => {
      seen.push({ url: String(url), init });
      return String(url).endsWith("/v1/jobs")
        ? response({ job_id: "job_1" })
        : response({ key: String(url).split("/").at(-1) });
    },
  });
  const attachments = [
    { key: "a.txt", content_type: "text/plain", kind: "document", body: "alpha" },
    { key: "b.wav", content_type: "audio/wav", kind: "voice", body: new Uint8Array([1, 2]) },
  ];

  await client.createJobWithAttachments(request, attachments);

  assert.equal(seen.length, 3);
  assert.deepEqual(
    seen.slice(1).map(({ url }) => url.split("/").at(-1)).sort(),
    ["a.txt", "b.wav"],
  );
});

test("multipart sends job first, preserves metadata and request headers", async () => {
  let captured;
  const client = new UbagClient({
    baseUrl: "https://gateway.example",
    appSecret: "test-secret",
    fetch: async (_url, init) => {
      captured = init;
      return response({ job_id: "job_2" });
    },
  });

  await client.createJobMultipart(
    request,
    [
      {
        key: "report.pdf",
        filename: "clinical-report.pdf",
        content_type: "application/pdf",
        kind: "document",
        body: new Uint8Array([37, 80, 68, 70]),
      },
      {
        key: "voice.webm",
        filename: "note.webm",
        content_type: "audio/webm",
        kind: "voice",
        body: new Uint8Array([26, 69, 223, 163]),
      },
    ],
    { idempotencyKey: "idem-attachments" },
  );

  assert.ok(captured.body instanceof FormData);
  assert.deepEqual([...captured.body.keys()], ["job", "report.pdf", "voice.webm"]);
  assert.equal(captured.body.get("report.pdf").type, "application/pdf");
  assert.equal(captured.body.get("report.pdf").name, "clinical-report.pdf");
  assert.equal(captured.body.get("voice.webm").type, "audio/webm");
  assert.equal(captured.headers.get("Authorization"), "Bearer test-secret");
  assert.equal(captured.headers.get("Ubag-Api-Version"), "2026-05-22");
  assert.equal(captured.headers.get("Idempotency-Key"), "idem-attachments");
  assert.equal(captured.headers.has("Content-Type"), false);
});

test("exports attachment hard limits", () => {
  assert.equal(UBAG_ATTACHMENT_MAX_FILE_BYTES, 32 * 1024 * 1024);
  assert.equal(UBAG_ATTACHMENT_MAX_MANIFEST_FILES, 32);
});

File: packages/sdk-go/attachments_test.go
package ubag

import (
	"context"
	"encoding/json"
	"io"
	"mime"
	"mime/multipart"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
)

func attachmentRequest() JSON {
	return JSON{
		"client": JSON{},
		"job": JSON{
			"target":       "mock",
			"command_type": "chat.prompt",
			"input":        JSON{"prompt": "inspect"},
		},
	}
}

func TestCreateJobWithAttachmentsUploadsEveryFile(t *testing.T) {
	var mu sync.Mutex
	var uploaded []string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/v1/jobs" {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusAccepted)
			_, _ = io.WriteString(w, `{"job_id":"job_1"}`)
			return
		}
		mu.Lock()
		uploaded = append(uploaded, r.URL.Path)
		mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, `{}`)
	}))
	defer server.Close()

	client, err := NewClient(server.URL, WithHTTPClient(server.Client()))
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.CreateJobWithAttachments(context.Background(), attachmentRequest(), []AttachmentUpload{
		{Key: "a.txt", ContentType: "text/plain", Kind: "document", Body: []byte("a")},
		{Key: "b.wav", ContentType: "audio/wav", Kind: "voice", Body: []byte("b")},
	})
	if err != nil {
		t.Fatal(err)
	}
	mu.Lock()
	defer mu.Unlock()
	if len(uploaded) != 2 ||
		!strings.Contains(strings.Join(uploaded, "|"), "/a.txt") ||
		!strings.Contains(strings.Join(uploaded, "|"), "/b.wav") {
		t.Fatalf("uploads = %#v", uploaded)
	}
}

func TestCreateJobMultipartPreservesOrderMetadataAndHeaders(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Idempotency-Key"); got != "idem-attachments" {
			t.Errorf("idempotency key = %q", got)
		}
		if got := r.Header.Get("Ubag-Api-Version"); got != DefaultAPIVersion {
			t.Errorf("api version = %q", got)
		}
		mediaType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
		if err != nil || mediaType != "multipart/form-data" {
			t.Fatalf("content type = %q, err=%v", mediaType, err)
		}
		reader := multipart.NewReader(r.Body, params["boundary"])
		var names []string
		var filenames []string
		var contentTypes []string
		for {
			part, err := reader.NextPart()
			if err == io.EOF {
				break
			}
			if err != nil {
				t.Fatal(err)
			}
			names = append(names, part.FormName())
			filenames = append(filenames, part.FileName())
			contentTypes = append(contentTypes, part.Header.Get("Content-Type"))
			_, _ = io.Copy(io.Discard, part)
		}
		if strings.Join(names, ",") != "job,report.pdf,voice.webm" {
			t.Errorf("part names = %#v", names)
		}
		if filenames[1] != "clinical-report.pdf" || filenames[2] != "note.webm" {
			t.Errorf("filenames = %#v", filenames)
		}
		if contentTypes[1] != "application/pdf" || contentTypes[2] != "audio/webm" {
			t.Errorf("content types = %#v", contentTypes)
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusAccepted)
		_ = json.NewEncoder(w).Encode(JSON{"job_id": "job_2"})
	}))
	defer server.Close()

	client, err := NewClient(server.URL, WithHTTPClient(server.Client()))
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.CreateJobMultipart(
		context.Background(),
		attachmentRequest(),
		[]AttachmentUpload{
			{Key: "report.pdf", Filename: "clinical-report.pdf", ContentType: "application/pdf", Kind: "document", Body: []byte("pdf")},
			{Key: "voice.webm", Filename: "note.webm", ContentType: "audio/webm", Kind: "voice", Body: []byte("webm")},
		},
		WithIdempotencyKey("idem-attachments"),
	)
	if err != nil {
		t.Fatal(err)
	}
}

func TestAttachmentLimits(t *testing.T) {
	if AttachmentMaxFileBytes != 32*1024*1024 || AttachmentMaxManifestFiles != 32 {
		t.Fatalf("limits = %d bytes, %d files", AttachmentMaxFileBytes, AttachmentMaxManifestFiles)
	}
}
